x/staking/src/types/query.rs

```diff
@@ -41,11 +54,122 @@ impl From<QueryValidatorRequest> for QueryValidatorRequestRaw {
 
 impl Protobuf<QueryValidatorRequestRaw> for QueryValidatorRequest {}
 
+/// QueryDelegationRequest is request type for the Query/Delegation RPC method.
+#[derive(Clone, Debug, PartialEq)]
+pub struct QueryDelegationRequest {
+    /// delegator_addr defines the delegator address to query for.
+    pub delegator_address: AccAddress,
+    /// validator_addr defines the validator address to query for.
+    pub validator_address: ValAddress,
+}
+
+impl TryFrom<QueryDelegationRequestRaw> for QueryDelegationRequest {
+    type Error = Error;
+
+    fn try_from(raw: QueryDelegationRequestRaw) -> Result<Self, Self::Error> {
+        let delegator_address = AccAddress::from_bech32(&raw.delegator_address)
+            .map_err(|e| Error::DecodeAddress(e.to_string()))?;
+        let validator_address = ValAddress::from_bech32(&raw.validator_address)
+            .map_err(|e| Error::DecodeAddress(e.to_string()))?;
+
+        Ok(QueryDelegationRequest {
+            delegator_address,
+            validator_address,
+        })
+    }
+}
+
+#[derive(Clone, PartialEq, Message)]
+pub struct QueryDelegationRequestRaw {
+    #[prost(string)]
+    pub delegator_address: String,
+    #[prost(string)]
+    pub validator_address: String,
+}
+
+impl From<QueryDelegationRequest> for QueryDelegationRequestRaw {
+    fn from(query: QueryDelegationRequest) -> QueryDelegationRequestRaw {
+        Self {
+            delegator_address: query.delegator_address.to_string(),
+            validator_address: query.validator_address.to_string(),
+        }
+    }
+}
+
+impl Protobuf<QueryDelegationRequestRaw> for QueryDelegationRequest {}
+
+/// QueryRedelegationRequest is request type for the Query/Redelegation RPC method.
+#[derive(Clone, Debug, PartialEq)]
+pub struct QueryRedelegationRequest {
+    /// delegator_addr defines the delegator address to query for.
+    pub delegator_address: Option<AccAddress>,
```

Comment:
Should these be required rather than optional?

Reply:
The struct can have type without `Option`. I used this signature that is close to sdk signature because it allows to have one node side handler for a few queries. [Here](https://github.com/rumos-io/gears/pull/105/files#diff-56b566899aff451a0f8dcccb07a4c0906dc4952ff18c9fdf17d5839adc234fdaR70) you can find match statement for the queries. 

If I'll change the type, then I need to have several additional structures for queries and additional handlers. It's just another solution. 

Should I change the code?

Reply:
If you can still handle a query (i.e. without returning an error) without a field, then it should be optional.
If I understand you correctly, you're re-using the struct for several queries. For some queries some fields are optional while for other queries they're required. Is that right? In which case go with whatever you prefer.

Reply:
>  For some queries some fields are optional while for other queries they're required. Is that right? In which case go with whatever you prefer.

Yes, it's right. The struct is used for several queries and the client code defines required/optional fields. I prefer to have one struct for redelegation queries because the other structs are the subsets of the current one.